File: examples/ios-harness/Sources/UmbrellaTestHarness/NativeBridges/AttestationBridge.swift
import Foundation
import DeviceCheck
import UmbrellaFFI

/// `AttestationBridge` — Apple App Attest (`DCAppAttestService`) обёртка
/// для генерации platform attestation token'а.
///
/// # Скоуп Блока 7.8
///
/// Skeleton-реализация: Rust `AttestationProvider` trait (см.
/// `crates/umbrella-client/src/attestation/provider_trait.rs`) **не
/// экспонирован** через uniffi как `callback_interface`. `AttestationBridge`
/// — самостоятельный iOS утилитарный класс; реальная интеграция Rust ↔ Swift
/// через uniffi callback interface в Блоке 7.10.
///
/// `generateKey` создаёт attestation key pair внутри Secure Enclave,
/// `attestKey(clientDataHash:)` делает attestation assertion (cryptographic
/// proof что приложение running на genuine Apple device). Server-side
/// verification — в Блоке 7.10.
///
/// # Block 7.8 scope
///
/// Skeleton only: the Rust `AttestationProvider` trait
/// (`crates/umbrella-client/src/attestation/provider_trait.rs`) is **not**
/// exposed via uniffi as a `callback_interface`. `AttestationBridge` is a
/// standalone iOS utility; real Rust ↔ Swift integration via a uniffi
/// callback interface lands in Block 7.10.
///
/// `generateKey` creates an attestation key pair inside Secure Enclave,
/// `attestKey(clientDataHash:)` produces an attestation assertion
/// (cryptographic proof that the app runs on a genuine Apple device). Server
/// verification is wired in Block 7.10.
final class AttestationBridge {
    enum BridgeError: Error {
        case notSupported
        case native(String)
    }

    private let service = DCAppAttestService.shared
    private var keyId: String?

    /// Генерирует attestation assertion для `serverNonce` (32-байтовый
    /// challenge от Sealed Server). Возвращает raw token bytes —
    /// переложенные в `PlatformAttestation` в Блоке 7.10.
    ///
    /// # Errors
    ///
    /// - `BridgeError.notSupported` если устройство не поддерживает App Attest.
    /// - `BridgeError.native` для underlying SDK error.
    ///
    /// Generates an attestation assertion for `serverNonce` (a 32-byte
    /// challenge from the Sealed Server). Returns the raw token bytes, which
    /// Block 7.10 repackages into `PlatformAttestation`.
    func freshToken(serverNonce: Data) async throws -> Data {
        guard service.isSupported else {
            throw BridgeError.notSupported
        }

        if keyId == nil {
            keyId = try await withCheckedThrowingContinuation { cont in
                service.generateKey { id, err in
                    if let err = err {
                        cont.resume(throwing: err)
                    } else if let id = id {
                        cont.resume(returning: id)
                    } else {
                        cont.resume(throwing: BridgeError.native("generateKey returned nil"))
                    }
                }
            }
        }

        let clientDataHash = serverNonce.subdata(in: 0..<min(32, serverNonce.count))

        return try await withCheckedThrowingContinuation { cont in
            service.attestKey(keyId!, clientDataHash: clientDataHash) { assertion, err in
                if let err = err {
                    cont.resume(throwing: err)
                } else if let assertion = assertion {
                    cont.resume(returning: assertion)
                } else {
                    cont.resume(throwing: BridgeError.native("attestKey returned nil"))
                }
            }
        }
    }

    /// Платформа — `iOS`. Matches Rust `Platform::iOs`.
    ///
    /// Platform — `iOS`. Matches Rust `Platform::iOs`.
    func platformTag() -> String { "iOs" }
}

File: examples/android-harness/app/src/main/java/xyz/umbrellax/testharness/nativebridges/AttestationBridge.kt
package xyz.umbrellax.testharness.nativebridges

import android.content.Context
import android.util.Base64
import com.google.android.play.core.integrity.IntegrityManagerFactory
import com.google.android.play.core.integrity.IntegrityTokenRequest
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * `AttestationBridge` — Google Play Integrity API wrapper для получения
 * platform attestation token'а.
 *
 * ## Скоуп Блока 7.9
 *
 * Skeleton: Rust `AttestationProvider` trait (см.
 * `crates/umbrella-client/src/attestation/provider_trait.rs`) **не
 * экспонирован** через uniffi как `callback_interface`. `AttestationBridge`
 * — standalone Kotlin утилитарный класс; real Rust ↔ Kotlin wiring
 * через uniffi callback interface в Блоке 7.10.
 *
 * `IntegrityManager.requestIntegrityToken(request)` производит cryptographic
 * proof что приложение running на genuine Android device с Play-installed APK.
 * Server-side verification — в Блоке 7.10.
 *
 * ## Block 7.9 scope
 *
 * Skeleton only: the Rust `AttestationProvider` trait is **not** exposed via
 * uniffi as a `callback_interface`. `AttestationBridge` is a standalone
 * Kotlin utility; real Rust ↔ Kotlin wiring via a uniffi callback interface
 * lands in Block 7.10.
 *
 * `IntegrityManager.requestIntegrityToken(request)` produces a cryptographic
 * proof that the app runs on a genuine Android device with a Play-installed
 * APK. Server-side verification is wired in Block 7.10.
 */
class AttestationBridge(private val ctx: Context) {

    private val manager = IntegrityManagerFactory.create(ctx)

    /** Platform tag. Matches Rust `Platform::Android`. */
    fun platformTag(): String = "Android"

    /**
     * Запрашивает Play Integrity token с `serverNonce` как challenge.
     *
     * Requests a Play Integrity token with `serverNonce` as the challenge.
     */
    suspend fun freshToken(serverNonce: ByteArray): ByteArray = suspendCancellableCoroutine { cont ->
        val nonceBase64 = Base64.encodeToString(
            serverNonce,
            Base64.NO_WRAP or Base64.URL_SAFE or Base64.NO_PADDING
        )
        val request = IntegrityTokenRequest.builder().setNonce(nonceBase64).build()

        manager.requestIntegrityToken(request)
            .addOnSuccessListener { response ->
                val token = response.token().toByteArray(Charsets.UTF_8)
                cont.resume(token)
            }
            .addOnFailureListener { err ->
                cont.resumeWithException(
                    RuntimeException("play integrity: ${err.message}", err)
                )
            }
    }
}
